### ISSUE TITLE
Have kubelet pass pod annotations directly to CNI plugins

### ISSUE THREAD

**Issue by leblancd** [kubernetes/kubernetes#69882]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:
Currently, if a CNI network plugin needs access to [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) that are configured on a pod, it needs to do a GET/read of the pod manifest from the Kubernetes API server, based on the pod's name.

**What you expected to happen**:
With the introduction of [CNI release 0.4.0](https://github.com/containernetworking/cni/blob/master/SPEC.md) (which supports CNI API version 0.2.0), it should be possible to make changes to kubelet such that pod annotations (and possibly labels) can be dynamically inserted into the CNI network configuration that is passed to CNI plugin binaries via stdin. In many cases, this would eliminate the need for CNI plugins to perform Kubernetes API reads of pod manifests.

This can be achieved by using the new [args field](https://github.com/containernetworking/cni/blob/master/CONVENTIONS.md#args-in-network-config) in the CNI network configuration. This field is intended to allow runtimes (e.g. kubelet) to dynamically insert per-container (per-pod, in the case of Kubernetes) configuration when a container/pod is being created.

The [CNI conventions for the args field](https://github.com/containernetworking/cni/blob/master/CONVENTIONS.md#args-in-network-config) currently suggest the use of a "labels" field to convey mappings to the plugin:
```
   "args":{  
      "cni":{  
         "labels": [{"key": "app", "value": "myapp"}]
      }
   },
```
It's not clear whether this "labels" field should be used ("overloaded") for Kubernetes pod annotations... maybe the CNI maintainers would prefer that we create a new "annotations" field in parallel with "labels".

The use of this feature would also require the use of CNI version 0.4.0 binaries, so we should consider bumping up the Kubernetes reference version for CNI from 0.3.0.

**How to reproduce it (as minimally and precisely as possible)**:
Refer to code [here](https://github.com/kubernetes/kubernetes/blob/release-1.10/pkg/kubelet/network/cni/cni.go#L285). This is approximately where changes would be needed.

**Anything else we need to know?**:
Use of this feature will require CNI versions 0.4.0 or newer. The Kubernetes reference version for CNI should be bumped from 0.3.0 to 0.4.0 at some point.

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by leblancd** [comment]
area /network

**Comment by leblancd** [comment]
sig /network

**Comment by danwinship** [comment]
/remove-lifecycle stale
/lifecycle frozen

Multus would make use of this too (cc @dcbw), to figure out if it needs to create additional networks for a pod.

And we could just pass the entire pod JSON. I think plenty of network plugins could make use of that. (HostPort pods are one thing in particular that the plugin might need to deal with that aren't labels or annotations.)

**Comment by kmala** [comment]
Do we have a priority for this? and i can work on this if needed

**Comment by danwinship** [comment]
I mean, we're surviving just fine without this feature, so I guess
/priority backlog
but it would be nice to have

**Comment by kmala** [comment]
@danwinship Thanks for confirming my assumption
This is kind of important for us as there is a usecase to write custom CNI plugins similar to existing meta plugins and it requires more information from POD than just the name and  namespace. 
Is it okay if i can implement this feature and create PR or do you think i should start discussion on the feature before implementation?

**Comment by danwinship** [comment]
Well, the discussion above is incomplete; no one ever decided exactly how the data would be passed, and it's not clear to me if you're talking about just passing the annotations or if you mean the entire pod json. So, if you want to describe what you're planning to implement before writing too much code, that might be good.

**Comment by ofiliz** [comment]
I am also interested in helping with defining the requirements and the specific mechanisms. I thought passing the entire pod json is probably too much, but maybe we can pass pod labels matching a well-known pattern ("cni*"?) as args to libcni?

Additionally, kubelet could call different CNI plugins (not just the same plugin with a different configuration) based on a well-known pod label. This would be useful for launching pods with different networking requirements on the same node.

In any case, I'm interested in discussing and helping with the implementation.

**Comment by kmala** [comment]
Thanks @danwinship and @ofiliz  
This is what i have in mind:
1) we should have naming convention such that we can pass only those annotations to the cni plugin. I am thinking `containernetwork.k8s.io` or `cni.k8s.io` as the prefix for the annotations
2) We can add those annotations as key value pairs to the args field  in addition to the existing pod name and namespace while creating the runtime conf, both cni and kubenet already have annotations available to them https://github.com/kubernetes/kubernetes/blob/d2eecfe2ef3785b272cb86638472716917ade35d/pkg/kubelet/dockershim/network/cni/cni.go#L365
https://github.com/kubernetes/kubernetes/blob/d2eecfe2ef3785b272cb86638472716917ade35d/pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go#L342

**Comment by danwinship** [comment]
>     1. we should have naming convention such that we can pass only those annotations to the cni plugin.

No, there are already other annotations that don't match that naming convention that plugins want access to. Eg, the [Network Attachment annotation](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#bookmark=id.qlmlur5ks48t) or the [pod bandwidth annotations](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/bandwidth/utils.go#L39). You should pass all of the annotations.

**Comment by kmala** [comment]
Thanks for giving more context. I will make a PR passing all the annotations.

**Comment by mars1024** [comment]
@danwinship @kmala  I don't think we need to pass annotations to CNI plugin, because it's CRI's job to translate pod spec or annotations to CNI RuntimeConfig.
For example, in different scenes, bandwidth may have different annotation keys, like bandwidth.k8s.io or bandwidth.cncf.io, actually bandwidth CNI plugin can't and shouldn't recognize both of them, so we should define some new CNI RuntimeConfigs and translate annotations to it, then pass them to CNI plugin, this will keep CNI plugin more general.

cc @dcbw @squeed

**Comment by khenidak** [comment]
I am concerned with the amount of information we pass into CNI (and how updated they are at the time of `Setup` call). Sure we can work around that by passing the entire pod spec. But do we ensure that pod status updates are fed into CNI as well?

I think we should keep the `setup` and `teardown` interfaces as simple as possible, maybe pass in a service account token (with prober default RBAC roles for it). If the CNI functionality depends pod spec (or other parts of the data that may be linked to that particular pod) that were not originally transmitted in `teardown` then it is up to the CNI runtime to read the api-server. 

The above represents a challenge to CNI developers because CNI is currently stateless, so data will have to be re-read as needed.

**Comment by danwinship** [comment]
@mars1024 This is about CNI plugins that are specifically interested in Kubernetes-specific things. There are already other bits of kubernetes-specific information that get passed to plugins (eg, the namespace of the pod). This would just be one more. CNI plugins that want to be container-platform-agnostic would just ignore the kubernetes-specific information

@khenidak No, we would not pass updates to the CNI plugin. Plugins that want to track the state of a Pod after creation time would have to continue doing that on their own, just like they do now. The goal here is just to simplify things for CNI plugins that need the pod annotations (or more) at `setup` time. eg, Multus (or any other "CNI Delegating Plugin" as defined in the [Kubernetes Network Custom Resource Definition De-facto Standard](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd)) needs to look at the pod annotations to see if it is supposed to only delegate to the primary network plugin, or if it needs to also call additional CNI plugins to create additional network interfaces. Currently it does this by making an apiserver request whenever it receives a `CNI_ADD` request, but if kubelet passed the pod annotations to it as part of the request, then it wouldn't have to do that.

**Comment by khenidak** [comment]
@danwinship i understand your point, thank you. 

The point i was trying to make. today we will add annotation, because they are needed. Later some CNI will be working with `labels`. Even later some other CNI will will need to work with `owner-ref` etc.. So i what i was trying to say is to find a process where the entire system is accessible `in a regulated fashion` to CNI.

**Comment by dcbw** [comment]
@khenidak @kmala @danwinship we've periodically discussed the issue in SIG Network in the past. And I think the reasons not to pass annotations still hold. Those are roughly that:

1) there can be *lots* of annotations, most of which are generic and the CNI plugin likely doesn't care about. Do we really want/need to push all of them to the plugin?

2) why are annotations special and things like the whole PodSpec or labels or other information not special? eg why stop at annotations?

3) plugins that need these annotations almost always already have a KubeClient anyway, so they can go retrieve the whole PodSpec if they want it. And they probably also need CRDs and stuff like that.

Basically we drew a line between plugins that are pretty dumb (eg flannel, bridge, kubenet, etc) and plugins that are more capable and already use KubeClients. And decided that it wasn't worth the downsides (complexity, size, processing, etc) to put annotations and/or other stuff into the CNI plugin. I don't think I've seen compelling arguments to revisit that decision :(

**Comment by kmala** [comment]
1) one of my first proposal is pass only specific annotations which have a domain prefix like `containernetwork.k8s.io` or `cni.k8s.io`. So, i am okay to make changes to push only few of them.
2) The reason for only annotations is because it is the preferred way of adding metadata to the pod. Though labels can also be used to add metadata, its mainly used for identifying the objects and selection. I am okay to even pass the whole podspec if we think it is useful for the plugins.
3) The plugin's can call apiserver to get the entire podspec but this would mean there would be an api call to apiserver for every add, delete event of pod and also much more if the CNI plugin needs information during handling the check method. We are already doing this and this is increasing a lot of load on the apiserver and decreasing the performance. 
I understand that the existing plugins doesn't need much metadata except the pod name and namespace name, but for people who are writing their own custom plugins for their internal networking and firewall restrictions, the handling would be different for different workloads(apps) and information is stored in metadata. The current way of calling kubernetes for each pod wouldn't scale in large clusters and if the CNI already supports passing of metadata using extra args then that should be the preferred way is what i think.

**Comment by ofiliz** [comment]
@dcbw I think while your point 3 above, "plugins that need these annotations already have a KubeClient anyway" is certainly true, this is only because we provide no other way for such plugins to achieve their goals. In complex Kubernetes networking setups there is now an explosion of daemons running on each node for this very reason. Common examples are daemons to configure sidecar proxies and network policy agents. These daemons consume node resources and increase deployment complexity. Every watch and get on API server also causes node scalability problems.

Another related problem we are solving here is not just what config to pass to CNI plugins, but also which CNI plugins to call. Existing design forces all plugins to be called for all pods. The common solution to that problem is to use a delegating CNI plugin, but because those plugins have no in-context information to decide which plugin(s) to call, they also have to fetch pod specs, inflating the problem.

**Comment by danwinship** [comment]
> "plugins that need these annotations already have a KubeClient anyway" is certainly true, this is only because we provide no other way for such plugins to achieve their goals.

But most of them are still going to need to need the KubeClient even if we provide them the annotations. eg:



> The common solution to that problem is to use a delegating CNI plugin, but because those plugins have no in-context information to decide which plugin(s) to call, they also have to fetch pod specs, inflating the problem.

But if the pod has the network attachment annotation, then that will point to a CRD which the delegating plugin will also need to look at. So it's still going to need a KubeClient anyway.

**Comment by kmala** [comment]
I understand that some plugins will/might still need to call KubeClient which we may not be able to solve but still there are multiple use-cases like custom firewall and network CNI plugins which would require the metadata of the pod and this feature would help them a lot.

**Comment by squeed** [comment]
For dynamic selection of network (e.g. Multus-style), I don't see any way how we can avoid a Kube-client in the near future. Mostly because of status reporting. Some day, we should probably standardize on a way to get multiple networks in the CRI, but I don't think that's going to happen now.

However, I'm more sympathetic to the allowing per-pod tunables. There is some precedent for this with the special bandwidth annotation, which is directly translated in to a CNI capability argument.  However, not all possible parameters should be capability args.

When the CNI was originally conceived, it was definitely expected that end-users would be able to pass runtime arguments directly to the CNI plugins. It was intended to look something like `rkt run --net=bridge:IP=1.2.3.4:FOO=BAR`. I think there is a precedent for exposing this mechanism.

So I would be in favor of adding a `cni.k8s.io/` annotation space. I think this is better than every single CNI binary doing it differently.

**Comment by danwinship** [comment]
OK, as discussed here and in the corresponding PR, there are issues with this idea/implementation:
- some people think it's wrong in general
- it's definitely wrong to do it only for dockershim and not for CRI
- confusion in the PR about exactly what the best way of passing the args would be

If we are going to do this, [it needs a KEP](https://github.com/kubernetes/enhancements/tree/master/keps) so we can hash all of that out (and get input from CRI people as well).
I'm going to close the PR but leave the issue open for now.

**Comment by reith** [comment]
I've been watching this issue since I liked our CNI to not watch for Pods but we already had kubeclient to manage CRDs representing our container links, so I was in agreement with @ofiliz . Now I'm changing our CNI for a nested environment, e.g. a Pod of a kubernetes in VMs which have their interfaces managed by another kubernetes. This inner CNI should talk with two API servers, inner kubernetes for Pods and upper kubernetes for Links of VMs. If I had Pod annotations from kubelet, I could make CNI to talk with just upper kubernets. It's still a narrow situation and relates to our design decisions, so probably not a requirement.  I just wanted to share a case that kubeclient is not really needed and some Pod information can suffice. Still how much information is needed is a subject of debate.

**Comment by tallclair** [comment]
Related issue: https://github.com/kubernetes/kubernetes/issues/84248 (node local apicache)

This seems like a strong use case for a node-local API. Basically, address the problems with CNI plugins needing to watch the apiserver, without needing to load everything into the CNI API.

**Comment by aojea** [comment]
I think this can be closed or renamed to "kubelet pass pod annotations to CRI", since dockershim removal there is no CNI code in Kubelet

**Comment by uablrek** [comment]
> kubelet pass pod annotations to CRI

Does the CRI's have a KubeClient usually? And read the pod spec for other purposes?

If so, this may be a pure CRI-plugin -> CNI-plugin issue.

**Comment by LeoLiuYan** [comment]
Ping @danwinship  @reith @tallclair .

Have we made any progress on this PR?

**Comment by aojea** [comment]
heh, this is seems implemented in the CRI and Kubelet ... since 2016 :upside_down_face: 

https://github.com/kubernetes/kubernetes/pull/30049

This is the CRI API that creates the Pod

https://github.com/kubernetes/cri-api/blob/278e3b3a55da3b872712c5ea4a7a148aa433c298/pkg/apis/services.go#L71

the config structure has label and annotations fields

https://github.com/kubernetes/cri-api/blob/278e3b3a55da3b872712c5ea4a7a148aa433c298/pkg/apis/runtime/v1/api.pb.go#L1297-L1314

and here is the code on that kubelet that creates the Pod, it copies the annotations and labels of the pod to the CRI structure

https://github.com/kubernetes/kubernetes/blob/8ccd03226f81573b91550f21c6b8d63f65e7c2f1/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L79-L92

https://github.com/kubernetes/kubernetes/blob/8ccd03226f81573b91550f21c6b8d63f65e7c2f1/pkg/kubelet/kuberuntime/labels.go#L75-L88

so,for the people demand to see those values on the CNI,  the ball now is is containerd/CRIO or the container runtimes to cascade those down

Or just follows Lars suggestion and use the CRI directly, it is available from the host

> If so, this may be a pure CRI-plugin -> CNI-plugin issue.

https://github.com/kubernetes/cri-api/blob/278e3b3a55da3b872712c5ea4a7a148aa433c298/pkg/apis/services.go#L79-L81

I can't see what else can be done in Kubernetes, it already offers the values and an API to access them

/close

**Comment by wreed4** [comment]
SO... after days of pouring through these issues and tracking down where repos have moved to..... I figured out... This already totally works.  

https://github.com/containerd/containerd/pull/5026

It seems to have been added in 2020.  

To pass in annotations, simply add the following to the cni configuration file.
```json
"capabilities": {
     "io.kubernetes.cri.pod-annotations": true
}
```
It will add all the annotations from the pod into the `runtimeConfiguration` section of the conf it passes in on stdin.  Documented here: https://github.com/containernetworking/cni/blob/main/CONVENTIONS.md
I can't find any documentation for this setting or this capability, but it works! :) Maybe use at your own risk??

stdin before
```json
{
  "cniVersion": "0.3.1",
  "debugFile": "/var/log/dummy2",
  "name": "k8s-pod-network",
  "prevResult": {
    "cniVersion": "0.3.1",
    "interfaces": [
      {
        "name": "cali5990130b379"
      }
    ],
    "ips": [
      {
        "version": "4",
        "address": "10.1.150.30/32"
      }
    ],
    "dns": {}
  },
  "type": "myplugin"
}
```
stdin after
```json
{
  "capabilities": {
    "io.kubernetes.cri.pod-annotations": true
  },
  "cniVersion": "0.3.1",
  "debugFile": "/var/log/dummy2",
  "name": "k8s-pod-network",
  "prevResult": {
    "cniVersion": "0.3.1",
    "dns": {},
    "interfaces": [
      {
        "name": "cali5990130b379"
      }
    ],
    "ips": [
      {
        "address": "10.1.150.30/32",
        "version": "4"
      }
    ]
  },
  "runtimeConfig": {
    "io.kubernetes.cri.pod-annotations": {
      "cni.projectcalico.org/podIP": "",
      "cni.projectcalico.org/podIPs": "",
      "kubectl.kubernetes.io/restartedAt": "2023-02-24T18:53:10-05:00",
      "kubernetes.io/config.seen": "2023-04-14T14:15:29.227406628-04:00",
      "kubernetes.io/config.source": "api",
      "my.awesome.annotation": "yay"
    }
  },
  "type": "myplugin"
}
```

**Comment by aojea** [comment]
> I can't find any documentation for this setting or this capability, but it works! :) Maybe use at your own risk??

it's unspecified AFAIK, so it will be project dependent

**Comment by yiya1989** [comment]
> SO... after days of pouring through these issues and tracking down where repos have moved to..... I figured out... This already totally works.
> 
> [containerd/containerd#5026](https://github.com/containerd/containerd/pull/5026)
> 
> It seems to have been added in 2020.
> 
> To pass in annotations, simply add the following to the cni configuration file.
> 
> ```json
> "capabilities": {
>      "io.kubernetes.cri.pod-annotations": true
> }
> ```
> 
> It will add all the annotations from the pod into the `runtimeConfiguration` section of the conf it passes in on stdin. Documented here: https://github.com/containernetworking/cni/blob/main/CONVENTIONS.md I can't find any documentation for this setting or this capability, but it works! :) Maybe use at your own risk??
> 
> stdin before
> 
> ```json
> {
>   "cniVersion": "0.3.1",
>   "debugFile": "/var/log/dummy2",
>   "name": "k8s-pod-network",
>   "prevResult": {
>     "cniVersion": "0.3.1",
>     "interfaces": [
>       {
>         "name": "cali5990130b379"
>       }
>     ],
>     "ips": [
>       {
>         "version": "4",
>         "address": "10.1.150.30/32"
>       }
>     ],
>     "dns": {}
>   },
>   "type": "myplugin"
> }
> ```
> 
> stdin after
> 
> ```json
> {
>   "capabilities": {
>     "io.kubernetes.cri.pod-annotations": true
>   },
>   "cniVersion": "0.3.1",
>   "debugFile": "/var/log/dummy2",
>   "name": "k8s-pod-network",
>   "prevResult": {
>     "cniVersion": "0.3.1",
>     "dns": {},
>     "interfaces": [
>       {
>         "name": "cali5990130b379"
>       }
>     ],
>     "ips": [
>       {
>         "address": "10.1.150.30/32",
>         "version": "4"
>       }
>     ]
>   },
>   "runtimeConfig": {
>     "io.kubernetes.cri.pod-annotations": {
>       "cni.projectcalico.org/podIP": "",
>       "cni.projectcalico.org/podIPs": "",
>       "kubectl.kubernetes.io/restartedAt": "2023-02-24T18:53:10-05:00",
>       "kubernetes.io/config.seen": "2023-04-14T14:15:29.227406628-04:00",
>       "kubernetes.io/config.source": "api",
>       "my.awesome.annotation": "yay"
>     }
>   },
>   "type": "myplugin"
> }
> ```

it's very usefullfor me, tks!

**Comment by MaoJianwei** [comment]
I need this too, when it will be release by kubernetes?

**Comment by aojea** [comment]
https://github.com/kubernetes/kubernetes/issues/69882#issuecomment-1228138339

**Comment by MaoJianwei** [comment]
I don't think this issue is resolved, because a CNI plugin cannot get any annotations about the pod/deployment at all, now.

at 1.28 version.

@aojea

**Comment by BenTheElder** [comment]
This thread is "Have kubelet pass pod annotations directly to CNI plugins"

kubelet does not talk to CNI plugins anymore, kubelet talks to CRI implementations (like CRI-O) which in turn use CNI, and already passes that info to CRI.

The CRI implementations could choose to pass this info, but that is not in scope for this repo. https://github.com/cri-o/cri-o/issues/7122 is an example of where this should be tracked. It appears containerd already supports this.
